### PR TITLE
Make gotthard work reliably when total number of DB stacks exceed 100.

### LINKED ISSUE #11: gotthard fails to search through all cloud formation stacks when the number is more than 100
It works for some clusters but not for the others. While examining the issue, I found it outputs exactly 100 stacks, failing to acquire the next chunk of stacks from the Amazon API (namely, cf.list_stacks).

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Fix #11

### CODE REVIEW

File: gotthard/cli.py [hunk 1]
```diff
@@ -307,13 +307,23 @@ def get_spilo_stacks(cf, spilo_cluster):
...
-        if not spilo_cluster or stack_name.split('-')[-1] == spilo_cluster:
-            desc = cf.describe_stacks(StackName=stack_name)['Stacks'][0]
-            for k, v in {i['Key']: i['Value'] for i in desc['Tags']}.items():
-                if k == 'SpiloCluster' and (not spilo_cluster or v == spilo_cluster):
-                    yield stack_name
+    next_token = None
+    while True:
+        # AWS emits output in chunks, loop until we examine all of them.
+        current_chunk = cf.list_stacks(StackStatusFilter=stack_status) if not next_token \
+            else cf.list_stacks(StackStatusFilter=stack_status, NextToken=next_token)
```
Comment (a1exsh) [inline gotthard/cli.py]: I think what you actually want is this: http://boto3.readthedocs.io/en/latest/reference/services/cloudformation.html?highlight=%22PageIterator%22#CloudFormation.Paginator.ListStacks

Example use: https://github.com/zalando-zmon/zmon-aws-agent/blob/master/zmon_aws_agent/aws.py#L157  Though I don't see `build_full_result()` anywhere in the docs, I expect that to be supported across all paginator implementations.

File: gotthard/cli.py [hunk 2]
```diff
@@ -224,12 +222,12 @@ def get_remote_host(remote_host, regions):
         logging.warning(ce.response['Error']['Code'])
         if ce.response['Error']['Code'] in ['ExpiredToken', 'RequestExpired']:
             raise ClickException('\nAWS credentials have expired.\n' +
-                                 'Use the "mai" command line tool to get a new temporary access key.')
+                                 'Use the "zaws login" command line tool to get a new temporary access key.')
```
Comment (avaczi) [inline gotthard/cli.py]: `zaws require`?

File: gotthard/cli.py [hunk 3]
```diff
@@ -245,7 +243,7 @@ def setup_tunnel(user, odd_host, remote_host, remote_port, tunnel_port):
                    '{}@{}'.format(user, odd_host),
                    '-N']
 
-    process = subprocess.Popen(ssh_command, preexec_fn = os.setpgrp)
+    process = subprocess.Popen(ssh_command, preexec_fn=os.setpgrp)
```
Comment (avaczi) [inline gotthard/cli.py:248]: why this change? :D

**Comment by a1exsh** [comment]
I was a little surprised that we don't list boto in `requirements.txt`, had to browse source code in order to find the actual version being used.  It's `boto3`.

File: gotthard/cli.py [hunk 4]
```diff
@@ -307,13 +305,23 @@ def get_spilo_stacks(cf, spilo_cluster):
...
 
-    for stack in cf.list_stacks(StackStatusFilter=stack_status)['StackSummaries']:
-        stack_name = stack['StackName']
-        if not spilo_cluster or stack_name.split('-')[-1] == spilo_cluster:
-            desc = cf.describe_stacks(StackName=stack_name)['Stacks'][0]
-            for k, v in {i['Key']: i['Value'] for i in desc['Tags']}.items():
-                if k == 'SpiloCluster' and (not spilo_cluster or v == spilo_cluster):
-                    yield stack_name
+    next_token = None
+    while True:
```
Comment (avaczi) [inline gotthard/cli.py]: Isn't there a more 'sparing' (ie. more direct, without listing so many stacks) to find the stack name?  That would automatically remove the problem being solved here.

File: gotthard/cli.py (hunk 3 above)
Comment (alexeyklyukin) [reply gotthard/cli.py:248]: PEP8: Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

File: gotthard/cli.py (hunk 2 above)
Comment (alexeyklyukin) [reply gotthard/cli.py]: Good point, didn't know it exist, thanks!

File: gotthard/cli.py (hunk 4 above)
Comment (alexeyklyukin) [reply gotthard/cli.py]: Not unless you hardcode the stack as spilo-{whatevername} (which might be not true for some other teams). 
There seem to be no way to partially match the name, see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html

Comment (alexeyklyukin) [reply gotthard/cli.py]: Actually, I've checked the performance of the ``describe_stacks`` with the hard-coded name template of spilo-{version}, and it didn't show any significant speed improvements over ``list_stacks``. In cases of either getting through stacks manually or describing a specific stack the requests took from 4s to 9s and timed out occasionally for an unspecified reason.

File: gotthard/cli.py [hunk 5]
```diff
@@ -307,7 +305,11 @@ def get_spilo_stacks(cf, spilo_cluster):
                     'UPDATE_ROLLBACK_IN_PROGRESS', 'UPDATE_ROLLBACK_FAILED',
                     'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS', 'UPDATE_ROLLBACK_COMPLETE']
 
-    for stack in cf.list_stacks(StackStatusFilter=stack_status)['StackSummaries']:
+    paginator = cf.get_paginator('list_stacks')
+    stacks = paginator.paginate(StackStatusFilter=stack_status,
+                                PaginationConfig = {'MaxItems': 10000}).build_full_result()['StackSummaries']
```
Comment (a1exsh) [inline gotthard/cli.py]: PEP8 :-p

Comment (a1exsh) [reply gotthard/cli.py]: Why explicit MaxItems?  Do you really expect more than 10k CF stacks?

Comment (alexeyklyukin) [reply gotthard/cli.py]: As far as I understood, MaxItems gives you the number of items you get on a single run, before you need to request the NextToken, so 10K is set to avoid calling this function multiple times.

File: gotthard/cli.py (hunk 3 above)
Comment (avaczi) [reply gotthard/cli.py:248]: bah.  Sounds very arbitrary :D

File: gotthard/cli.py (hunk 5 above)
Comment (a1exsh) [reply gotthard/cli.py]: Why do you need the paginator then?  In case it's still over 10k?

File: gotthard/cli.py (hunk 4 above)
Comment (avaczi) [reply gotthard/cli.py]: I was trying to talk basically about if it is possible to filter the stack list _before_ receiving it, thus avoiding the whole paging issue and complicated code

File: gotthard/cli.py (hunk 5 above)
Comment (alexeyklyukin) [reply gotthard/cli.py]: On re-reading it, it's probably PageSize rather than MaxItems that I need.

Comment (alexeyklyukin) [reply gotthard/cli.py]: Well, with Paginators I can explicitly set the page size to be more than 100 items. I cannot find an explicit guarantee there that there will be only one page, as build_full_result returns the NextToken in the source code and there is no documentation.

File: gotthard/cli.py (hunk 3 above)
Comment (alexeyklyukin) [reply gotthard/cli.py:248]: Why?

File: gotthard/cli.py (hunk 4 above)
Comment (alexeyklyukin) [reply gotthard/cli.py]: That's exactly what describe_stacks do.

File: gotthard/cli.py [hunk 6]
```diff
@@ -307,7 +305,11 @@ def get_spilo_stacks(cf, spilo_cluster):
                     'UPDATE_ROLLBACK_IN_PROGRESS', 'UPDATE_ROLLBACK_FAILED',
                     'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS', 'UPDATE_ROLLBACK_COMPLETE']
 
-    for stack in cf.list_stacks(StackStatusFilter=stack_status)['StackSummaries']:
+    paginator = cf.get_paginator('list_stacks')
+    stacks = paginator.paginate(StackStatusFilter=stack_status,
+                                PaginationConfig = {'PageSize': 10000}).build_full_result()['StackSummaries']
```
Comment (a1exsh) [inline gotthard/cli.py]: This still suffers from PEP8 problem...

**Comment by a1exsh** [comment]
:+1:

**Comment by alexeyklyukin** [comment]
👍
